opwen_email_server/api/store_written_client_emails.py

```diff
@@ -20,33 +17,27 @@
         container=constants.CONTAINER_CLIENT_PACKAGES,
         provider=config.STORAGE_PROVIDER))
 
+logger = LogMixin()
 
-class _WrittenStorer(LogMixin):
-    def __call__(self, resource_id: str):
-        emails = STORAGE.fetch_objects(resource_id)
+@celery.task
```

Comment:
It's often a good idea to keep the integration boundaries between your own code and a third party system as small as possible. This makes it easier to test your own code (because you don't have to set up the machinery of the third party system) and also makes it easier to evolve your code since the third party system integration can easily be swapped out.

What do I concretely mean by this? I often like to keep my business logic (e.g. storing a resource, sending an email, etc.) as simple Python functions and then have framework wrappers that call through to these functions, e.g. here we have our web framework call simple functions since we use Connexion, but if we used Flask we could still have the same structure with the `@app.route('...')` definitions calling through to the business logic functions. This means that we can test business logic without having to spin up a web server.

We should do something similar for the Celery tasks. I.e., keep all the business logic code (such as _WrittenStorer) as they are and create a single integration file that lists all the tasks, a bit like this:

```py
from celery import Celery
from opwen_email_server.api import store_written_client_emails

celery = Celery()

@celery.task
def store(resource_id):
  store_written_client_emails.store(resource_id)
```

This approach has a number of benefits:
- Everything that touches Celery is contained to one file which makes it very easy to see all the background tasks of the project.
- We don't have to point Celery to particular modules where to discover tasks since everything is self-contained.
- It's easy to swap out the queue framework in the future: only replace one file and the rest of the project doesn't have to change since it's not even aware that Celery exists.
- Fewer code changes for the PR.

You can find an example of this pattern in a recent project of that I worked on: [faceanalysis/tasks.py](https://github.com/CatalystCode/faceanalysis/blob/fb3ebc9b18b79ad927627f3039ae8b6e27c64f62/app/faceanalysis/tasks.py#L12).

Reply:
I was actually thinking of this. I tried to sneak celery in there by changing as little as possible. I should have asked more questions about the structure of Celery and its respective tasks. Thanks for the example as well!